flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/AbstractDispatcherLeaderProcess.java

```diff
@@ -0,0 +1,254 @@
+		log.info("Stopping {}.", getClass().getSimpleName());
+
+		final CompletableFuture<Void> dispatcherServiceTerminationFuture = closeDispatcherService();
+
+		final CompletableFuture<Void> onCloseTerminationFuture = FutureUtils.composeAfterwards(
+			dispatcherServiceTerminationFuture,
+			this::onClose);
+
+		FutureUtils.forward(
+			onCloseTerminationFuture,
+			this.terminationFuture);
+
+		state = State.STOPPED;
+	}
+
+	private CompletableFuture<Void> closeDispatcherService() {
+		if (dispatcherService != null) {
+			return dispatcherService.closeAsync();
+		} else {
+			return FutureUtils.completedVoidFuture();
+		}
+	}
+
+	protected abstract void onStart();
+
+	protected CompletableFuture<Void> onClose() {
+		return FutureUtils.completedVoidFuture();
+	}
+
+	final void completeDispatcherSetup(DispatcherService dispatcherService) {
+		runIfStateIs(
+			State.RUNNING,
+			() -> completeDispatcherSetupInternal(dispatcherService));
+	}
+
+	private void completeDispatcherSetupInternal(DispatcherService createdDispatcherService) {
+		Preconditions.checkState(dispatcherService == null, "The DispatcherService can only be set once.");
+		dispatcherService = createdDispatcherService;
+		dispatcherGatewayFuture.complete(createdDispatcherService.getGateway());
+		FutureUtils.forward(createdDispatcherService.getShutDownFuture(), shutDownFuture);
+	}
+
+	final <V> Optional<V> supplyUnsynchronizedIfRunning(Supplier<V> supplier) {
+		synchronized (lock) {
+			if (state != State.RUNNING) {
+				return Optional.empty();
+			}
+		}
+
+		return Optional.of(supplier.get());
+	}
+
+	final <V> Optional<V> supplyIfRunning(Supplier<V> supplier) {
+		synchronized (lock) {
+			if (state != State.RUNNING) {
+				return Optional.empty();
+			}
+
+			return Optional.of(supplier.get());
+		}
+	}
+
+	final void runIfStateIs(State expectedState, Runnable action) {
+		runIfState(expectedState::equals, action);
+	}
+
+	private void runIfStateIsNot(State notExpectedState, Runnable action) {
+		runIfState(
+			state -> !notExpectedState.equals(state),
+			action);
+	}
+
+	private void runIfState(Predicate<State> actionPredicate, Runnable action) {
+		synchronized (lock) {
+			if (actionPredicate.test(state)) {
+				action.run();
+			}
+		}
+	}
+
+	final <T> Void onErrorIfRunning(T ignored, Throwable throwable) {
+		synchronized (lock) {
+			if (state != State.RUNNING) {
+				return null;
+			}
+		}
+
+		if (throwable != null) {
+			closeAsync();
+			fatalErrorHandler.onFatalError(throwable);
+		}
+
+		return null;
+	}
+
+	protected enum State {
+		CREATED,
+		RUNNING,
+		STOPPED
+	}
+
+	// ------------------------------------------------------------
+	// Internal classes
+	// ------------------------------------------------------------
+
+	interface DispatcherServiceFactory {
+		DispatcherService create(
+			DispatcherId fencingToken,
+			Collection<JobGraph> recoveredJobs,
+			JobGraphWriter jobGraphWriter);
+	}
+
+	interface DispatcherService extends AutoCloseableAsync {
```

Comment:
Shall we call this one `DispatcherContainer` since it is a thin wrapper of

```
Dispatcher dispatcher;
DispatcherGateway dispatcherGateway;
```

and its methods don't like a service.

What is more important is that I'm afraid it is a bit confusing the different between this `DispatcherService` and `DispatcherServices` which are totally different thing but share a similar name under the same scope.

Reply:
I see your concern Tison. However, I'm not sure whether I like `DispatcherContainer` too much. My reasoning was a bit along the lines of `Executor` and `ExecutorService` where the `XService` gives you access to lifecycle methods.

Reply:
What about `DispatcherGatewayService`?

Reply:
Make sense. Thanks for the update.